crates/mir/src/lower/function.rs

```diff
@@ -946,28 +944,29 @@ impl<'db, 'a> BodyLowerHelper<'db, 'a> {
             }
 
             AnalyzerCallType::ValueMethod { method, .. } => {
+                let resolved_generics = self.resolve_generics_args(method, &args);
+
                 let mut method_args = vec![self.lower_method_receiver(func)];
                 method_args.append(&mut args);
 
                 let func_id = if method.is_generic(self.db.upcast()) {
-                    let concrete_args = method_args
-                        .iter()
-                        .map(|val| {
-                            self.builder
-                                .value_ty(*val)
-                                .analyzer_ty(self.db)
-                                .expect("invalid parameter")
-                        })
-                        .collect::<Vec<_>>();
                     self.db
-                        .mir_lowered_monomorphized_func_signature(*method, concrete_args)
+                        .mir_lowered_monomorphized_func_signature(*method, resolved_generics)
                 } else {
                     self.db.mir_lowered_func_signature(*method)
                 };
 
                 self.builder
                     .call(func_id, method_args, CallType::Internal, source)
             }
+            AnalyzerCallType::TraitValueMethod {
+                trait_id, method, ..
+            } if trait_id.is_std_trait(self.db.upcast(), EMITTABLE_TRAIT_NAME)
+                && method.name(self.db.upcast()) == EMIT_FN_NAME =>
+            {
+                let event = self.lower_method_receiver(func);
+                self.builder.emit(event, source)
+            }
```

Comment:
This is `#event/#indexed` specific.

Reply:
It is, but so is the status quo with the `Event` type and the `lower_event_type` function. After all, this PR doesn't aim to change the internal handling of events it only changes the trigger from being `emit SomeEventType` to being trait based via  `Emittable.emit(..)`.

Reply:
> this PR doesn't aim to change the internal handling of events

This is the point. Again, I'd like to ask how long this stopgap solution exist. And I think the core of the codebase is affected too much just to remove the superficial syntax.

Reply:
>Again, I'd like to ask how long this stopgap solution will remain

I don't know but I would assume it is still a while out until we can implement the actual event logic in Fe code natively.

>And I think the core of the codebase is affected too much

But wouldn't you agree that the PR actually lowered the complexity of the code base in general? I mean, look at all the event handling logic that was deleted. More code was deleted than added in this PR. It's actually very little code that was added to enable the struct + trait based event handling.

Reply:
I don't think complexity could/should be evaluated by the lines of code added/deleted. And how extent the codebase is affected is almost nothing to do with the LoC added/deleted.

I think the way of abstraction introduced here makes the codebase more complicated.

> There is no ImplId because there is no impl block for it.

This inconsistency didn't exist before the PR, for example. As I mentioned, this would break the invariant of the trait-related methods should provide, and if the PR is merged as is, it will force us to care about the `Emittable` trait wherever we handle traits. You exposed the name of the `Emittable` trait as a public constant, and it proves that we should treat the trait specially even now, and in the future, we would start writing other stopgaps to adapt to the unexpected behavior of the core functions. So stopgaps will be chained and expanded.

Also, this PR "hides" the stopgap behind the pseudo abstraction. This makes it difficult to remove these stopgaps when we finally reach the "right" solution. I mean Rust compiler won't complain about the stopgaps in the form of compile errors. So, as a result, we'll have to find and remove them manually. I am not confident in remembering the exact functions and locations where stopgaps are embedded. And as I said, these stopgaps will be chained/expanded as time passes.

So, to be honest, I have to say this PR increases the complexity of the code, and that's why I strongly disagree with merging this PR as is.

Reply:
No worries, we can try to work out a solution when we co-work in Berlin next week. And if we can not find a satisfying solution than I can live with the current system based on `event` and `emit` until we have all the ingredients to fully implement something like the `Emittable` trait in Fe itself.

That said, I do disagree on the question of complexity. I think having the `event` type and the `emit` keyword with all of its extra checks is more complex than having a trait in the std lib that is automagically implemented by the compiler as a temporary solution. Especially since it is also ensured that no one else can try to implement it manually for the time being.

Reply:
I don't disagree with adding the `Emittable` trait and implementing it to some types, though I don't think it should be implemented to *all* types. What I disagree with is the way of realizing it. Again, I still think we shouldn't arbitrarily tweak the core functions on which many other parts depend.

But yeah, we will be able to discuss this face-to-face the next week, so it'd be fine to stop the discussion on GH.